File: LeetCode/202.快乐数.cpp
/*
 * @lc app=leetcode.cn id=202 lang=cpp
 *
 * [202] 快乐数
 */
#include <bits/stdc++.h>
using namespace std;

class Solution {
  public:
    int d[10] = {0, 1, 4, 9, 16, 25, 36, 49, 64, 81};
    bool isHappy(int n) {
        set<int> s;
        while (true) {
            if (n == 1)
                return true;
            if (s.count(n) == 1)
                return false;
            s.insert(n);
            int t = 0;
            while (n) {
                t += d[n % 10];
                n /= 10;
            }
            n = t;
        }
    }
};

File: LeetCode/Python/202.快乐数.py
#
# @lc app=leetcode.cn id=202 lang=python3
#
# [202] 快乐数
#


# @lc code=start
class Solution:
    def isHappy(self, n: int) -> bool:
        def getNext(number: int) -> int:
            sum = 0
            while number > 0:
                number, digit = divmod(number, 10)
                sum += digit**2
            return sum

        slow, fast = n, getNext(n)
        while fast != 1 and slow != fast:
            slow = getNext(slow)
            fast = getNext(getNext(fast))
        return fast == 1


# @lc code=end
